compiler/src/dotty/tools/dotc/transform/CheckUnused.scala

```diff
@@ -375,9 +375,16 @@ class CheckUnused private (phaseMode: PhaseMode, suffix: String) extends MiniPha
     val ctxs = ctx.outersIterator
     while !done && ctxs.hasNext do
       val cur = ctxs.next()
-      if cur.owner eq sym then
-        addCached(cachePoint, Definition)
-        return // found enclosing definition
+      if cur.owner.userSymbol eq sym then
+        if sym.is(Package) then
+          if cur.isSimplePackage then
+            precedence = Definition
+            importer = null
+          // else package p.q is not available as a simple name but it might be imported
+          end if
+          done = true
+        else
+          return // found enclosing definition, don't register the reference
```

Comment:
This special handling isn't necessary. Just set a flag for "found enclosing" (so don't register reference), then in outer context it will find the member (for "simple package" or other defs) and not find the "hidden" dotted package name. The latter case won't terminate right away (but that could be fixed).

Reply:
Also I already told tpolecat that I deleted the `return`.